internal/integrations/integrations.go

```diff
@@ -0,0 +1,138 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package integrations // import "github.com/elastic/opentelemetry-collector-components/internal/integrations"
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/pipeline"
+)
+
+// ErrNotFound is the error returned when an integration cannot be found.
+var ErrNotFound = errors.New("not found")
+
+// Integration is the interface for integrations that can be resolved as configuration.
+type Integration interface {
+	// Resolve resolves the parametrizable integration. It uses the params to replace placeholders
+	// in the integration, and removes or ignores everything not referenced by the indicated
+	// pipelines.
+	Resolve(ctx context.Context, params map[string]any, pipelines []pipeline.ID) (*Config, error)
+}
+
+// Finder is the interface for extensions that can be used to find integrations by their names.
+type Finder interface {
+	FindIntegration(ctx context.Context, name string) (Integration, error)
+}
+
+// Host is the interface a host must implement to be used to locate finders along extensions.
+type Host interface {
+	// GetExtensions returns the map of extensions. Only enabled and created extensions will be returned.
+	// GetExtensions can be called by the component anytime after Component.Start() begins and
+	// until Component.Shutdown() ends.
+	GetExtensions() map[component.ID]component.Component
+}
+
+// Find looks for integrations in extensions of the host that implement the IntegrationFinder interface.
+func Find(ctx context.Context, logger *zap.Logger, host Host, name string) (Integration, error) {
```

Comment:
Should the user be able to differentiate between no integration matches the given `name` vs there are no integrations at all? So far all return paths return `ErrNotFound` which let the user not differentiate in an automated way.

Reply:
Yes, for users the error mentions that no extension is found.

I have changed it to a named `ErrNotConfigured` error, but still wrapping `ErrNotFound`, so components can just check if the integration is not found, while they propagate the error.